### ISSUE TITLE
Data stored in test instance objects is never released

### ISSUE THREAD

**Issue by bmerry** [pytest-dev/pytest#11374]
## Description

I have a test suite that started life in nosetests, and a lot of the tests still use xunit style with a `setup_method` that creates fixture objects and stores them in the instance of the test class. I ran into a problem with tests exhausting resources, and it seems like pytest is hanging on to references to the instances of the test classes, preventing these fixture objects from being garbage collected.

Here's a proof-of-concept:

conftest.py:
```python
import gc

def pytest_runtest_logfinish(nodeid, location):
    import test_leak
    gc.collect()
    print(f"\nExpensive objects: {test_leak.ExpensiveObject.count}", end="")
```

test_leak.py:
```python
import gc
from typing import ClassVar

import pytest


class ExpensiveObject:
    count: ClassVar[int] = 0

    def __init__(self) -> None:
        ExpensiveObject.count += 1

    def __del__(self) -> None:
        ExpensiveObject.count -= 1


class TestClass:
    def setup_method(self) -> None:
        self.obj = ExpensiveObject()

    @pytest.mark.parametrize("x", range(5))
    def test(self, x: int) -> None:
        pass
```

Output:
```
test_leak.py::TestClass::test[0] PASSED                                                                                                   [ 20%]
Expensive objects: 1
test_leak.py::TestClass::test[1] PASSED                                                                                                   [ 40%]
Expensive objects: 2
test_leak.py::TestClass::test[2] PASSED                                                                                                   [ 60%]
Expensive objects: 3
test_leak.py::TestClass::test[3] PASSED                                                                                                   [ 80%]
Expensive objects: 4
test_leak.py::TestClass::test[4] PASSED                                                                                                   [100%]
Expensive objects: 5
```

So despite the calls to `gc.collect()`, the number of ExpensiveObjects is increasing. In contrast, if I use a `pytest.fixture`, they are all cleaned up immediately.

## Environment
```
Package        Version
-------------- -------
exceptiongroup 1.1.3
iniconfig      2.0.0
packaging      23.1
pip            23.2.1
pluggy         1.3.0
pytest         7.4.0
setuptools     68.1.2
tomli          2.0.1
wheel          0.41.2
```

- Pytest version: 7.4.0
- OS version: Ubuntu 22.04
- Python version: 3.10.12

- [x] a detailed description of the bug or problem you are having
- [x] output of `pip list` from the virtual environment you are using
- [x] pytest and operating system versions
- [x] minimal example if possible

**Comment by bluetech** [comment]
>  In contrast, if I use a pytest.fixture, they are all cleaned up immediately.

`setup_method` is translated to a fixture so that's surprising. Can you show the fixture alternative which does get cleaned up?

**Comment by RonnyPfannschmidt** [comment]
It's normal python behavior, fixture references eventually go, instance variables don't

As there's no actual resource cleanup, this is a user error

**Comment by bmerry** [comment]
> > In contrast, if I use a pytest.fixture, they are all cleaned up immediately.
> 
> `setup_method` is translated to a fixture so that's surprising. Can you show the fixture alternative which does get cleaned up?

Sure.

```python
import gc
from typing import ClassVar

import pytest


class ExpensiveObject:
    count: ClassVar[int] = 0

    def __init__(self) -> None:
        ExpensiveObject.count += 1

    def __del__(self) -> None:
        ExpensiveObject.count -= 1


class TestWithFixture:
    def obj(self) -> ExpensiveObject:
        return ExpensiveObject()

    @pytest.mark.parametrize("x", range(5))
    def test(self, x: int) -> None:
        pass
```

Output:
```
test_leak.py::TestWithFixture::test[0] PASSED
Expensive objects: 0
test_leak.py::TestWithFixture::test[1] PASSED
Expensive objects: 0
test_leak.py::TestWithFixture::test[2] PASSED
Expensive objects: 0
test_leak.py::TestWithFixture::test[3] PASSED
Expensive objects: 0
test_leak.py::TestWithFixture::test[4] PASSED
Expensive objects: 0
```

> It's normal python behavior, fixture references eventually go, instance variables don't

It's normal if pytest is holding references to the instances of TestClass for all the tests. Is there a reason it needs to hold references to those instances once the test has finished?

**Comment by bmerry** [comment]
It looks like the unittest.TestCase support also behaves fairly well compared to `setup_method`:

```python
import gc
import unittest
from typing import ClassVar

import pytest


class ExpensiveObject:
    count: ClassVar[int] = 0

    def __init__(self) -> None:
        ExpensiveObject.count += 1

    def __del__(self) -> None:
        ExpensiveObject.count -= 1


class TestUnittest(unittest.TestCase):
    def setUp(self) -> None:
        self.obj = ExpensiveObject()

    def test1(self) -> None:
        pass

    def test2(self) -> None:
        pass

    def test3(self) -> None:
        pass
```

Output:
```
test_leak.py::TestUnittest::test1 PASSED
Expensive objects: 1
test_leak.py::TestUnittest::test2 PASSED
Expensive objects: 1
test_leak.py::TestUnittest::test3 PASSED
Expensive objects: 0
```

**Comment by RonnyPfannschmidt** [comment]
Indeed, that's Part of why a number of fixture and parametrization Features are not available there as unittest support uses unittest machinery instead of pytest machinery

**Comment by bluetech** [comment]
> > setup_method is translated to a fixture so that's surprising. Can you show the fixture alternative which does get cleaned up?
>
> Sure.

There is no fixture in your example; maybe you mean to make `obj` a fixture and request it in `test`?

**Comment by bmerry** [comment]
> > > setup_method is translated to a fixture so that's surprising. Can you show the fixture alternative which does get cleaned up?
> > 
> > 
> > Sure.
> 
> There is no fixture in your example; maybe you mean to make `obj` a fixture and request it in `test`?

Whoops, problem between keyboard and chair! Here's a corrected version; the output is unchanged.

```python
import gc
from typing import ClassVar

import pytest


class ExpensiveObject:
    count: ClassVar[int] = 0

    def __init__(self) -> None:
        ExpensiveObject.count += 1

    def __del__(self) -> None:
        ExpensiveObject.count -= 1


class TestWithFixture:
    @pytest.fixture
    def obj(self) -> ExpensiveObject:
        return ExpensiveObject()

    @pytest.mark.parametrize("x", range(5))
    def test(self, x: int, obj: ExpensiveObject) -> None:
        pass
```

**Comment by bluetech** [comment]
OK so as @RonnyPfannschmidt said, pytest (python) items keep a reference to their underlying object, which in the case of test methods is the method bound to a fresh class instance, and that keeps the class instance alive for the duration of the session.

Although I agree this is a gotcha with class tests, I don't think this is something that is going to change for now (would require introducing a lifecycle for the `item.obj`s which is not straightforward), so you should do one of the following:

- Switch to pytest fixtures
- Manually cleanup in teardown

**Comment by bmerry** [comment]
Thanks for that clue. Adding the following hook seems to work for me, but I can appreciate that it might not work universally. Maybe it's worth mentioning in the documentation?

```python
@pytest.hookimpl(trylast=True)
def pytest_runtest_teardown(item, nextitem):
    if inspect.ismethod(item.obj):
        item.obj.__self__.__dict__.clear()
```

**Comment by RonnyPfannschmidt** [comment]
There may be a need for extra resource cleanup hints,

However gc based resource cleanup should never be suggested as its a well documented Bad practice

It's strongly recommended that fixtures that's create resources also clean them up

It's also universally a bad idea to use dunder del unless you implemented a low-level Type

**Comment by bmerry** [comment]
> However gc based resource cleanup should never be suggested as its a well documented Bad practice

I'm not sure I follow. Given:

```python
class Foo:
    def setup_method(self):
        self.squares = [i * i for i in range(100000000)]
```

how do you suggest cleaning up the resource (memory) used by that list without involving the garbage collector?

```python
    def teardown_method(self):
        del self.squares
```

**Comment by RonnyPfannschmidt** [comment]
@bmerry to me it appears you are confusing resources and memory there

what i wanted to get at is that something like `ExpensiveObject` shouldnt not use `__del__` to clean up but use a contextmanager

so that code using it would use

```python
with expensive_allocation() as expensive_object:
   expensive_object.do_work()
```

the oversimplified example for squares would look more like

```python
@pytest.fixture
def squares():
   squares = [i * i for i in range(100000000)]
   yield squares
   squares.clear() # here we `invalidate`  the contents for all objects that may still refer to it by accident
   ```

**Comment by bmerry** [comment]
> @bmerry to me it appears you are confusing resources and memory there

"Resource" is a very general word with different meaning in different contexts. It sounds like you're using it to mean things like handles that are finite and should be closed. Ok, I'll stick with that definition to avoid confusion. As you say, the user should ideally release those (with context management or calling `.close()` or similar). And I confess that for the specific case where I ran into this issue, it is file handles that I've run out of, and it's my own fault for making design choices that make it very difficult to provide a context manager (*).

However, that doesn't mean that other users might not be creating objects in `setup_method` that consume lots of memory (ala my toy `squares` examples), and are just assuming that the GC will be able to reclaim that memory after the test, and will be surprised to find that they need to explicitly drop references to avoid pytest using O(n_tests * avg_test_memory) RAM rather than O(n_tests + max_test_memory).
